### ISSUE TITLE
Suggestion: Use Block Skip Instead of Attention Skip in STG Integration

### ISSUE THREAD

**Issue by kinam0252** [Lightricks/LTX-Video#111]
Hi, I'm one of the authors of [STG](https://github.com/junhahyung/STGuidance). First off, thanks for integrating STG into LTXVideo!

I wanted to suggest a small change—recently, we found that **Block Skip** gives **much better performance improvements** than **Attention Skip** in LTXVideo. 

Would you be open to updating the current attention residual code to use block residual instead? 
You can refer to our **implementation** [here](https://github.com/junhahyung/STGuidance/blob/main/LTXVideo/pipeline_stg_ltx.py).

Really appreciate the great work on this video model!

https://github.com/user-attachments/assets/3f23d5c1-d1ce-4e4c-a6f8-2e03c3a70150

https://github.com/user-attachments/assets/a0e190cc-36fd-409f-8955-5c27bad2a26c

**Comment by yoavhacohen** [comment]
Hi @kinam0252 thanks for your comment and for implementing STG for LTX-Video.

I believe that we support both options:

**Attention skip**
https://github.com/Lightricks/LTX-Video/blob/caf9f0c4670b462b51c845abff7f5731ed138364/ltx_video/models/transformers/attention.py#L1062

**Block skip:** 
https://github.com/Lightricks/LTX-Video/blob/caf9f0c4670b462b51c845abff7f5731ed138364/ltx_video/models/transformers/attention.py#L1088

One can choose between them here:
https://github.com/Lightricks/LTX-Video/blob/caf9f0c4670b462b51c845abff7f5731ed138364/inference.py#L351

Currently, `SkipLayerStrategy.Attention` is the default, but based on your suggestion, we will definitely evaluate both options and set the better-performing one as the default.

Our implementation differs from yours to allow for efficient compilation. Skipping is handled using masks so that everything runs in a batch. Could you review our implementation to ensure we haven’t missed anything and that it’s functionally equivalent to yours?

Thanks again for your insights!

**Comment by kinam0252** [comment]
Thanks for the clarification! My bad for not explaining it clearly in the first place.  

Your current implementation is correct, but what I was referring to is a **different kind of Block Skip**. 
Right now, the Block Skip in your code skips **Attention Block**, but what we found to be more effective is skipping entire **Transformer Block** instead.  

To implement this, we modified the `forward` of the Transformer Block to `forward_with_stg`, as shown below:  

```python
def forward_with_stg(
        self,
        hidden_states: torch.Tensor,
        encoder_hidden_states: torch.Tensor,
        temb: torch.Tensor,
        image_rotary_emb: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
        encoder_attention_mask: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
    
        hidden_states_ptb = hidden_states[2:]
        encoder_hidden_states_ptb = encoder_hidden_states[2:]
        ...
        hidden_states[2:] = hidden_states_ptb
        encoder_hidden_states[2:] = encoder_hidden_states_ptb
    
        return hidden_states

if self.do_spatio_temporal_guidance:
    for i in stg_applied_layers_idx:
        self.transformer.transformer_blocks[i].forward = types.MethodType(forward_with_stg, self.transformer.transformer_blocks[i])
```
This change improves performance in our experiments. Thanks again for your work on this!

**Comment by yoavhacohen** [comment]
Thanks for clarifying, I think I understand now.

An equivalent implementation to yours would be to add the following lines just before the end of `BasicTransformerForward.forward(...)`:
```
        if skip_layer_mask is not None and skip_layer_strategy == SkipLayerStrategy.TransformerBlock:
            hidden_states = hidden_states * skip_layer_mask + original_hidden_states * (1.0 - skip_layer_mask)
```

And define:
```
original_hidden_states = hidden_states
```

This way, we would effectively skip the entire Transformer block when `SkipLayerStrategy.TransformerBlock` is selected, while still keeping everything in the same batch to allow for compilation efficiency.

Can you confirm that this matches your intended behavior?

**Comment by kinam0252** [comment]
Yes, it looks good to me!
